failure in ud/test_ucp_sockaddr.onesided_disconnect_c2s/3

http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/14415/label=hpc-test-node-gpu,worker=2/console

```
17:42:18 [ RUN      ] ud/test_ucp_sockaddr.onesided_disconnect_c2s/3 <ud_v>
17:42:18 [     INFO ] Testing 11.151.71.0:39055
17:42:28 /scrap/jenkins/workspace/hpc-ucx-pr-3/label/hpc-test-node-gpu/worker/2/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:470: Failure
17:42:28 Value of: status
17:42:28   Actual: 1
17:42:28 Expected: UCS_OK
17:42:28 Which is: 0
17:42:28 /scrap/jenkins/workspace/hpc-ucx-pr-3/label/hpc-test-node-gpu/worker/2/contrib/../test/gtest/common/test.cc:269: Failure
17:42:28 Failed
17:42:28 Got 1 errors during the test
17:42:28 /scrap/jenkins/workspace/hpc-ucx-pr-3/label/hpc-test-node-gpu/worker/2/contrib/../test/gtest/common/test.cc:273: Failure
17:42:28 Failed
17:42:28 Got 1 warnings during the test
17:42:28 
[  FAILED  ] ud/test_ucp_sockaddr.onesided_disconnect_c2s/3, where GetParam() = ud_v (10089 ms)
```

repro:

`taskset -c 0,1 make -C test/gtest test GTEST_FILTER=dc_ud/test_ucp_sockaddr.concurrent_disconnect_c2s/4 UCX_LOG_LEVEL_TRIGGER=error GTEST_REPEAT=100000`
```
Repeating all tests (iteration 7385) . . .

Note: Google Test filter = dc_ud/test_ucp_sockaddr.concurrent_disconnect_c2s/4
Note: Randomizing tests' orders with a seed of 50560 .
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from dc_ud/test_ucp_sockaddr
[ RUN      ] dc_ud/test_ucp_sockaddr.concurrent_disconnect_c2s/4 <dc_x,ud_v,ud_x,mm>
[     INFO ] Testing 65.65.65.12:42827
[r-vmb-ppc-jenkins:29916:0:28670]  ucp_worker.c:538  error 'Connection reset by remote peer' will not be handled for ep 0x3fff932c0000 - / since no error callback is installed
[r-vmb-ppc-jenkins:29916:0:28670] Process frozen, press Enter to attach a debugger...
```

backtrace for thread which prints the error:
```
#0  0x00003fff9ac03548 in read () from /usr/lib64/libpthread.so.0
(gdb) bt
#0  0x00003fff9ac03548 in read () from /usr/lib64/libpthread.so.0
#1  0x00003fff9afec114 in ucs_error_freeze (
    message=0x3fff996ed960 "error 'Connection reset by remote peer' will not be handled for ep 0x3fff932c0000 - / since no error callback is installed") at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/debug/debug.c:899
#2  ucs_handle_error (
    message=0x3fff996ed960 "error 'Connection reset by remote peer' will not be handled for ep 0x3fff932c0000 - / since no error callback is installed") at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/debug/debug.c:1071
#3  0x00003fff9afe71f8 in ucs_fatal_error_message (
    file=0x3fff9aebb060 "/labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucp/core/ucp_worker.c", 
    line=<optimized out>, function=<optimized out>, 
    message_buf=0x3fff996ed960 "error 'Connection reset by remote peer' will not be handled for ep 0x3fff932c0000 - / since no error callback is installed")
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/debug/assert.c:33
#4  0x00003fff9afed660 in ucs_log_default_handler (
    file=0x3fff9aebb060 "/labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucp/core/ucp_worker.c", 
    line=<optimized out>, function=0x3fff9aeba8f0 <__func__.14816> "ucp_worker_set_ep_failed", 
    level=<optimized out>, format=<optimized out>, ap=0x3fff996ede88 "\210\061\022\233\377?")
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/debug/log.c:141
#5  0x00003fff9afed850 in ucs_log_dispatch (
    file=0x3fff9aebb060 "/labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucp/core/ucp_worker.c", 
    line=<optimized out>, function=0x3fff9aeba8f0 <__func__.14816> "ucp_worker_set_ep_failed", 
    level=<optimized out>, 
    format=0x3fff9aebb250 "error '%s' will not be handled for ep %p - %s/%s since no error callback is installed")
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/debug/log.c:192
#6  0x00003fff9ae4dfa8 in ucp_worker_set_ep_failed (worker=0x100151248d0, ucp_ep=0x3fff932c0000, 
    uct_ep=<optimized out>, lane=<optimized out>, status=<optimized out>)
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucp/core/ucp_worker.c:535
#7  0x00003fff9aea9004 in ucp_cm_server_connect_cb (ep=<optimized out>, arg=<optimized out>, 
    status=<optimized out>) at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucp/wireup/wireup_cm.c:741
#8  0x00003fff9af45ea4 in uct_cm_ep_server_connect_cb (cep=<optimized out>, status=<optimized out>)
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/uct/base/uct_cm.c:63
#9  0x00003fff9a52b3ec in uct_rdmacm_cm_ep_server_connect_cb (cep=<optimized out>, status=<optimized out>)
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/uct/ib/rdmacm/rdmacm_cm_ep.c:45
#10 0x00003fff9a52b468 in uct_rdmacm_cm_ep_error_cb (cep=<optimized out>, remote_data=<optimized out>, 
    status=<optimized out>) at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/uct/ib/rdmacm/rdmacm_cm_ep.c:70
#11 0x00003fff9a529dd4 in uct_rdmacm_cm_handle_event_disconnected (event=0x3fff8c0008e0, event=0x3fff8c0008e0)
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/uct/ib/rdmacm/rdmacm_cm.c:307
---Type <return> to continue, or q <return> to quit---
#12 uct_rdmacm_cm_process_event (cm=0x10014f40b30, event=0x3fff8c0008e0)
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/uct/ib/rdmacm/rdmacm_cm.c:388
#13 0x00003fff9a52a504 in uct_rdmacm_cm_event_handler (fd=<optimized out>, arg=0x10014f40b30)
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/uct/ib/rdmacm/rdmacm_cm.c:438
#14 0x00003fff9afcf72c in ucs_async_handler_invoke (handler=0x10014f25a50)
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/async/async.c:224
#15 ucs_async_handler_dispatch (handler=0x10014f25a50)
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/async/async.c:242
#16 ucs_async_dispatch_handlers (events=0x3fff996ee450, count=1)
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/async/async.c:271
#17 0x00003fff9afd4664 in ucs_async_thread_ev_handler (callback_data=<optimized out>, event=<optimized out>, 
    arg=0x3fff996ee6b0) at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/async/thread.c:86
#18 0x00003fff9aff8c8c in ucs_event_set_wait (event_set=0x10014ffcfc0, num_events=0x3fff996ee6c0, 
    timeout_ms=<optimized out>, event_set_handler=0x3fff9afd4600 <ucs_async_thread_ev_handler>, 
    arg=0x3fff996ee6b0) at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/sys/event_set.c:213
#19 0x00003fff9afd44c4 in ucs_async_thread_func (arg=0x100150c4ab0)
    at /labhome/yosefe/workspace/hpc/ucx/contrib/../src/ucs/async/thread.c:127
#20 0x00003fff9abf8af4 in start_thread () from /usr/lib64/libpthread.so.0
#21 0x00003fff9a764ef4 in clone () from /usr/lib64/libc.so.6
(gdb) p ucp_ep
$2 = (struct ucp_ep *) 0x3fff932c0000
(gdb) get_flags ucp_ep->flags
$3 = 3                                                                                                                
$4 = 4                                                                                                                         
$5 = 7                                                                                                                         
$6 = 10                                                                                                                        
$7 = 22 
(gdb) p worker->ep_config[ucp_ep->cfg_index].key.err_mode 
$11 = UCP_ERR_HANDLING_MODE_PEER
(gdb) p * (ucp_ep_ext_gen_t *) ( (char*)ucp_ep + 128*1024 )
$15 = {dest_ep_ptr = 70366919262208, user_data = 0x1001508b260, ep_list = {prev = 0x100151249f8, 
    next = 0x100151249f8}, err_cb = 0x0, {ep_match = {dest_uuid = 1099863050848, list = {prev = 0x3fff932e0028, 
        next = 0x0}}, flush_state = {reqs = {head = 0x10014f24a60, ptail = 0x3fff932e0028}, send_sn = 0, 
      cmpl_sn = 0}, listener = 0x10014f24a60, close_req = {req = 0x10014f24a60}}}
```

--- a/src/ucp/core/ucp_ep.c
+++ b/src/ucp/core/ucp_ep.c
@@ -766,6 +766,7 @@ void ucp_ep_cleanup_lanes(ucp_ep_h ep)
     }
 }
 
+/* Must be called with async lock held */
 void ucp_ep_disconnected(ucp_ep_h ep, int force)
 {
     /* remove pending slow-path progress in case it wasn't removed yet */
@@ -799,12 +800,15 @@ void ucp_ep_disconnected(ucp_ep_h ep, int force)
 
 unsigned ucp_ep_local_disconnect_progress(void *arg)
 {
-    ucp_request_t *req = arg;
+    ucp_request_t *req         = arg;
+    ucp_ep_h ep                = req->send.ep;
+    ucs_async_context_t *async = &ep->worker->async; /* ep becomes invalid */
 
     ucs_assert(!(req->flags & UCP_REQUEST_FLAG_COMPLETED));
 
-    ucp_ep_disconnected(req->send.ep, req->send.flush.uct_flags &
-                                      UCT_FLUSH_FLAG_CANCEL);
+    UCS_ASYNC_BLOCK(async);
+    ucp_ep_disconnected(ep, req->send.flush.uct_flags & UCT_FLUSH_FLAG_CANCEL);
+    UCS_ASYNC_UNBLOCK(async);
 
     /* Complete send request from here, to avoid releasing the request while
      * slow-path element is still pending */

--- a/src/ucp/core/ucp_worker.c
+++ b/src/ucp/core/ucp_worker.c
@@ -475,8 +475,14 @@ int ucp_worker_err_handle_remove_filter(const ucs_callbackq_elem_t *elem,
 {
     ucp_worker_err_handle_arg_t *err_handle_arg = elem->arg;
 
-    return (elem->cb == ucp_worker_iface_err_handle_progress) &&
-           (err_handle_arg->ucp_ep == arg);
+    if ((elem->cb == ucp_worker_iface_err_handle_progress) &&
+        (err_handle_arg->ucp_ep == arg)) {
+        /* release err handling argument to avoid memory leak */
+        ucs_free(err_handle_arg);
+        return 1;
+    }
+
+    return 0;
 }
 
 ucs_status_t ucp_worker_set_ep_failed(ucp_worker_h worker, ucp_ep_h ucp_ep,

--- a/test/gtest/common/test_helpers.h
+++ b/test/gtest/common/test_helpers.h
@@ -766,6 +766,12 @@ private:
 };
 
 
+template <typename T>
+static void deleter(T *ptr) {
+    delete ptr;
+}
+
+
 extern int    perf_retry_count;
 extern double perf_retry_interval;
 

--- a/test/gtest/ucp/test_ucp_sockaddr.cc
+++ b/test/gtest/ucp/test_ucp_sockaddr.cc
@@ -180,7 +180,8 @@ public:
     void start_listener(ucp_test_base::entity::listen_cb_type_t cb_type,
                         const struct sockaddr* addr, size_t addrlen)
     {
-        ucs_status_t status = receiver().listen(cb_type, addr, addrlen);
+        ucs_status_t status = receiver().listen(cb_type, addr, addrlen,
+                                                get_ep_params());
         if (status == UCS_ERR_UNREACHABLE) {
             UCS_TEST_SKIP_R("cannot listen to " + ucs::sockaddr_to_str(addr));
         }
@@ -494,10 +495,15 @@ public:
         /* The current expected errors are only from the err_handle test
          * and from transports where the worker address is too long but ud/ud_x
          * are not present, or ud/ud_x are present but their addresses are too
-         * long as well */
+         * long as well, in addition we can get disconnect events during test
+         * teardown.
+         */
         switch (status) {
         case UCS_ERR_REJECTED:
         case UCS_ERR_UNREACHABLE:
+        case UCS_ERR_CONNECTION_RESET:
+            UCS_TEST_MESSAGE << "ignoring error " <<ucs_status_string(status)
+                             << " on endpoint " << ep;
             return;
         default:
             UCS_TEST_ABORT("Error: " << ucs_status_string(status));
@@ -652,7 +658,8 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle) {
     ucs::sock_addr_storage listen_addr(test_addr.to_ucs_sock_addr());
     ucs_status_t status = receiver().listen(cb_type(),
                                             listen_addr.get_sock_addr_ptr(),
-                                            listen_addr.get_addr_size());
+                                            listen_addr.get_addr_size(),
+                                            get_ep_params());
     if (status == UCS_ERR_UNREACHABLE) {
         UCS_TEST_SKIP_R("cannot listen to " + ucs::sockaddr_to_str(&listen_addr));
     }

--- a/test/gtest/ucp/ucp_test.cc
+++ b/test/gtest/ucp/ucp_test.cc
@@ -459,14 +459,14 @@ void ucp_test_base::entity::connect(const entity* other,
 }
 
 ucp_ep_h ucp_test_base::entity::accept(ucp_worker_h worker,
-                                       ucp_conn_request_h conn_request,
-                                       const void *ep_user_data)
+                                       ucp_conn_request_h conn_request)
 {
+    ucp_ep_params_t ep_params = *m_server_ep_params;
     ucp_ep_h        ep;
-    ucp_ep_params_t ep_params;
-    ep_params.field_mask   = UCP_EP_PARAM_FIELD_USER_DATA |
-                             UCP_EP_PARAM_FIELD_CONN_REQUEST;
-    ep_params.user_data    = (void *)ep_user_data;
+
+    ep_params.field_mask  |= UCP_EP_PARAM_FIELD_CONN_REQUEST |
+                             UCP_EP_PARAM_FIELD_USER_DATA;
+    ep_params.user_data    = reinterpret_cast<void*>(this);
     ep_params.conn_request = conn_request;
 
     ucs_status_t status    = ucp_ep_create(worker, &ep_params, &ep);
@@ -502,6 +502,16 @@ void ucp_test_base::entity::empty_send_completion(void *r, ucs_status_t status) 
 void ucp_test_base::entity::accept_ep_cb(ucp_ep_h ep, void *arg) {
     entity *self = reinterpret_cast<entity*>(arg);
     int worker_index = 0; /* TODO pass worker index in arg */
+
+    /* take error handler from test fixture and add user data */
+    ucp_ep_params_t ep_params = *self->m_server_ep_params;
+    ep_params.field_mask &= UCP_EP_PARAM_FIELD_ERR_HANDLER;
+    ep_params.field_mask |= UCP_EP_PARAM_FIELD_USER_DATA;
+    ep_params.user_data   = reinterpret_cast<void*>(self);
+
+    void *req = ucp_ep_modify_nb(ep, &ep_params);
+    ASSERT_UCS_PTR_OK(req); /* don't expect this operation to block */
+
     self->set_ep(ep, worker_index, self->get_num_eps(worker_index));
 }
 
@@ -571,7 +581,9 @@ ucp_ep_h ucp_test_base::entity::revoke_ep(int worker_index, int ep_index) const 
 
 ucs_status_t ucp_test_base::entity::listen(listen_cb_type_t cb_type,
                                            const struct sockaddr* saddr,
-                                           socklen_t addrlen, int worker_index)
+                                           socklen_t addrlen,
+                                           const ucp_ep_params_t& ep_params,
+                                           int worker_index)
 {
     ucp_listener_params_t params;
     ucp_listener_h        listener;
@@ -599,6 +611,8 @@ ucs_status_t ucp_test_base::entity::listen(listen_cb_type_t cb_type,
     default:
         UCS_TEST_ABORT("invalid test parameter");
     }
+
+    m_server_ep_params.reset(new ucp_ep_params_t(ep_params));
 
     ucs_status_t status;
     {
@@ -644,7 +658,7 @@ unsigned ucp_test_base::entity::progress(int worker_index)
     if (!m_conn_reqs.empty()) {
         ucp_conn_request_h conn_req = m_conn_reqs.back();
         m_conn_reqs.pop();
-        ucp_ep_h ep = accept(ucp_worker, conn_req, this);
+        ucp_ep_h ep = accept(ucp_worker, conn_req);
         set_ep(ep, worker_index, std::numeric_limits<int>::max());
         ++progress_count;
     }

--- a/test/gtest/ucp/ucp_test.cc
+++ b/test/gtest/ucp/ucp_test.cc
@@ -612,7 +612,8 @@ ucs_status_t ucp_test_base::entity::listen(listen_cb_type_t cb_type,
         UCS_TEST_ABORT("invalid test parameter");
     }
 
-    m_server_ep_params.reset(new ucp_ep_params_t(ep_params));
+    m_server_ep_params.reset(new ucp_ep_params_t(ep_params),
+                             ucs::deleter<ucp_ep_params_t>);
 
     ucs_status_t status;
     {

--- a/test/gtest/ucp/ucp_test.h
+++ b/test/gtest/ucp/ucp_test.h
@@ -17,7 +17,6 @@
 
 #include <common/test.h>
 
-#include <memory>
 #include <queue>
 
 #define MT_TEST_NUM_THREADS       4
@@ -122,7 +121,7 @@ public:
         std::queue<ucp_conn_request_h>  m_conn_reqs;
         size_t                          m_err_cntr;
         size_t                          m_rejected_cntr;
-        std::auto_ptr<ucp_ep_params_t>  m_server_ep_params;
+        ucs::handle<ucp_ep_params_t*>   m_server_ep_params;
 
     private:
         static void empty_send_completion(void *r, ucs_status_t status);

--- a/test/gtest/ucp/ucp_test.h
+++ b/test/gtest/ucp/ucp_test.h
@@ -17,6 +17,7 @@
 
 #include <common/test.h>
 
+#include <memory>
 #include <queue>
 
 #define MT_TEST_NUM_THREADS       4
@@ -63,8 +64,7 @@ public:
         void connect(const entity* other, const ucp_ep_params_t& ep_params,
                      int ep_idx = 0, int do_set_ep = 1);
 
-        ucp_ep_h accept(ucp_worker_h worker, ucp_conn_request_h conn_request,
-                        const void *ep_user_data);
+        ucp_ep_h accept(ucp_worker_h worker, ucp_conn_request_h conn_request);
 
         void* modify_ep(const ucp_ep_params_t& ep_params, int worker_idx = 0,
                        int ep_idx = 0);
@@ -82,6 +82,7 @@ public:
 
         ucs_status_t listen(listen_cb_type_t cb_type,
                             const struct sockaddr *saddr, socklen_t addrlen,
+                            const ucp_ep_params_t& ep_params,
                             int worker_index = 0);
 
         ucp_ep_h ep(int worker_index = 0, int ep_index = 0) const;
@@ -121,6 +122,7 @@ public:
         std::queue<ucp_conn_request_h>  m_conn_reqs;
         size_t                          m_err_cntr;
         size_t                          m_rejected_cntr;
+        std::auto_ptr<ucp_ep_params_t>  m_server_ep_params;
 
     private:
         static void empty_send_completion(void *r, ucs_status_t status);

